File: providers/adzuna.py
"""Adzuna live job provider.

Adzuna (https://developer.adzuna.com/) offers a free jobs API that aggregates
listings from many boards (Indeed, Totaljobs, Reed, LinkedIn, CV-Library and
more). It maps cleanly onto PartFinder's needs: postcode location, radius in
miles, part-time filter, minimum salary, sort-by-date and a max-age window.

When credentials are absent or the request fails, ``search`` logs and returns an
empty list so the app gracefully falls back to the seed-backed providers.
"""

from __future__ import annotations

import logging
from typing import List, Optional

import requests

from config import CATEGORY_BY_KEY
from models import Job, SearchCriteria
from providers.base import JobProvider
from utils import format_pay, parse_iso_date

logger = logging.getLogger(__name__)

_BASE_URL = "https://api.adzuna.com/v1/api/jobs/{country}/search/1"


class AdzunaProvider(JobProvider):
    """Query the Adzuna API and normalize results into :class:`Job` objects."""

    name = "Adzuna"
    reliability = 0.92  # Aggregator with structured, reasonably clean data.

    def search(self, criteria: SearchCriteria) -> List[Job]:
        if not self.config.adzuna_configured:
            logger.info("Adzuna credentials not set; skipping live provider.")
            return []

        params = self._build_params(criteria)
        url = _BASE_URL.format(country=self.config.adzuna_country)
        try:
            response = requests.get(
                url,
                params=params,
                headers={"Accept": "application/json"},
                timeout=self.config.provider_timeout_seconds,
            )
            response.raise_for_status()
            payload = response.json()
        except (requests.RequestException, ValueError) as exc:
            logger.warning("Adzuna request failed: %s", exc)
            return []

        jobs = [self._to_job(item, criteria) for item in payload.get("results", [])]
        return self._stamp([job for job in jobs if job is not None])

    def _build_params(self, criteria: SearchCriteria) -> dict:
        """Translate search criteria into Adzuna query parameters."""
        category = CATEGORY_BY_KEY.get(criteria.category)
        what = " ".join(category.keywords) if category else criteria.category

        params = {
            "app_id": self.config.adzuna_app_id,
            "app_key": self.config.adzuna_app_key,
            "results_per_page": self.config.results_per_provider,
            "what": what,
            "where": self.config.origin_postcode,
            "distance": criteria.radius,  # miles
            "max_days_old": self.config.max_job_age_days,
            "sort_by": "date",
            "part_time": 1,  # Prefer part-time roles.
        }
        if category and category.adzuna_category:
            params["category"] = category.adzuna_category
        if category and category.contract:
            params["contract"] = 1
        return params

    def _to_job(self, item: dict, criteria: SearchCriteria) -> Optional[Job]:
        """Normalize one Adzuna result item into a :class:`Job`."""
        title = item.get("title")
        url = item.get("redirect_url")
        if not title or not url:
            return None

        posted = parse_iso_date(item.get("created", ""))
        if posted is None:
            return None

        salary_min = item.get("salary_min")
        salary_max = item.get("salary_max")
        location = item.get("location", {}) or {}
        contract_time = (item.get("contract_time") or "").replace("_", "-")

        return Job(
            title=title,
            company=(item.get("company", {}) or {}).get("display_name", "Unknown"),
            location=location.get("display_name", self.config.origin_postcode),
            pay=format_pay(salary_min, salary_max),
            employment_type=contract_time.title() or "Part-time",
            posted=posted,
            source=self.name,
            url=url,
            description=item.get("description", ""),
            latitude=item.get("latitude"),
            longitude=item.get("longitude"),
            category=criteria.category,
            salary_min=salary_min,
            salary_max=salary_max,
        )

File: web/js/adzuna.js
/*
 * PartFinder — Adzuna live job fetch (browser).
 * Builds the request and normalizes results into job objects with REAL
 * `redirect_url` Apply links. Mirrors providers/adzuna.py from the Flask app.
 */

const ADZUNA_BASE = "https://api.adzuna.com/v1/api/jobs/{country}/search/1";

function adzunaConfigured() {
    const c = window.PF_CONFIG || {};
    return Boolean(c.ADZUNA_APP_ID && c.ADZUNA_APP_KEY);
}

function buildAdzunaUrl(criteria) {
    const c = window.PF_CONFIG || {};
    const category = CATEGORY_BY_KEY[criteria.category];
    const what = category ? category.keywords.join(" ") : criteria.category;

    const params = new URLSearchParams({
        app_id: c.ADZUNA_APP_ID,
        app_key: c.ADZUNA_APP_KEY,
        results_per_page: "50",
        what: what,
        where: ORIGIN.postcode,
        distance: String(criteria.radius),
        max_days_old: String(MAX_JOB_AGE_DAYS),
        sort_by: "date",
        part_time: "1",
        "content-type": "application/json",
    });
    if (category && category.adzuna) params.set("category", category.adzuna);
    if (category && category.contract) params.set("contract", "1");

    const base = ADZUNA_BASE.replace("{country}", c.ADZUNA_COUNTRY || "gb");
    return `${base}?${params.toString()}`;
}

/*
 * Adzuna does not send CORS headers, so a browser blocks the direct call.
 * When PF_CONFIG.CORS_PROXY is set, route the request through it. Default is a
 * free, keyless proxy that returns the raw JSON with permissive CORS headers.
 */
function withCorsProxy(adzunaUrl) {
    const c = window.PF_CONFIG || {};
    const proxy = c.CORS_PROXY;
    if (!proxy) return adzunaUrl; // direct call (only where CORS is allowed)
    return proxy + encodeURIComponent(adzunaUrl);
}

/* Normalize one Adzuna result into a job object. */
function normalizeAdzuna(item, criteria) {
    if (!item || !item.title || !item.redirect_url) return null;
    const contractTime = (item.contract_time || "").replace("_", "-");
    return {
        title: item.title,
        company: (item.company && item.company.display_name) || "Unknown",
        location: (item.location && item.location.display_name) || ORIGIN.location,
        latitude: item.latitude ?? null,
        longitude: item.longitude ?? null,
        salary_min: item.salary_min ?? null,
        salary_max: item.salary_max ?? null,
        pay: formatPay(item.salary_min, item.salary_max),
        employment_type: contractTime ? contractTime.replace(/\b\w/g, (m) => m.toUpperCase()) : "Part-time",
        created: item.created || null,
        source: "Adzuna",
        url: item.redirect_url,
        description: item.description || "",
        category: criteria.category,
        distance: null,
    };
}

/*
 * Fetch live jobs. Resolves to { jobs } or throws an Error with a `.reason`
 * of "no-key" | "cors" | "http" | "network" so the UI can show the right state.
 */
async function fetchAdzunaJobs(criteria) {
    if (!adzunaConfigured()) {
        const e = new Error("Adzuna credentials not set.");
        e.reason = "no-key";
        throw e;
    }

    let response;
    try {
        const url = withCorsProxy(buildAdzunaUrl(criteria));
        response = await fetch(url, { headers: { Accept: "application/json" } });
    } catch (err) {
        // A blocked CORS request (or an unreachable proxy) surfaces here as a
        // TypeError ("Failed to fetch").
        const e = new Error("Could not reach Adzuna from the browser.");
        e.reason = "cors";
        throw e;
    }

    if (!response.ok) {
        const e = new Error(`Adzuna returned HTTP ${response.status}.`);
        e.reason = response.status === 401 || response.status === 403 ? "no-key" : "http";
        throw e;
    }

    const payload = await response.json();
    const jobs = (payload.results || [])
        .map((item) => normalizeAdzuna(item, criteria))
        .filter(Boolean);
    return { jobs };
}
